### PR TITLE
Fix User Menus+ handling of all user links and mentions (closes #1331)

### LINKED ISSUE #1331: Profiler (and possibly other extensions) broken by User Menus+ putting 'undefined' in place of usernames
While trying to use the Profiler extension, information for the blog undefined.tumblr.com would come up instead of the blog I was trying to find out about.

image of the Profiler extension, showing undefined.tumblr.com instead of the target blog

I looked into the code for this extension, and the problem seems to come in User Menus+ (or show_more in the extension editor). There are some kinds of links to blogs, such as reblog headers and mentions, which are not correctly updating the information used by User Menus+ when they're moused over.

---
Profiler gets the URL of the blog the user is asking about from a `data-url` attribute on its User Menus+ menu entry ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/profiler.js#L175)):

```javascript
menu_clicked: function(e) {

	var m_object = $(e.target);

	if (!m_object.hasClass("xkit-profiler")) {

		while (!m_object.hasClass("xkit-profiler")) {
			m_object = m_object.parent();
		}

	}

	/*...*/

	var user_url = $(m_object).attr('data-url');

	XKit.extensions.profiler.show(user_url);

},
```
This `data-url` attribute is defined in the function Profiler uses to add its User Menus+ entry ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/profiler.js#L147)):

```javascript
XKit.extensions.show_more.add_custom_menu("profiler", function(data) {
	var user_url = data.name;

	$(document).off("click", ".xkit-profiler-button-" + user_url, XKit.extensions.profiler.menu_clicked);
	$(document).on("click", ".xkit-profiler-button-" + user_url, XKit.extensions.profiler.menu_clicked);

	return "<div data-url=\"" + user_url + "\" class=\"xkit-profiler-button-" + user_url + " xkit-profiler\">Profiler</div>";
});
```

with a similar function for non-Classic menus. This depends on the `data` object passed to this function having the correct information.

Going over to User Menus+, we find the `data` that gets passed to this function is defined by an object called `popup_data` ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L572), likewise [here](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L215) for Classic menus).

```javascript
if (XKit.extensions.show_more.custom_menu_extension.length >= 0) {

	var m_data = XKit.extensions.show_more.popup_data;

	for (var i = 0; i < XKit.extensions.show_more.custom_menu_extension.length; i++) {

		var returned_menu = "";

		try {
			returned_menu = XKit.extensions.show_more.custom_menu_function[i](m_data);
		} catch (e) {
			returned_menu = "";
		}

		m_html = m_html + returned_menu;
	}

}
```

`popup_data` is updated by two functions `store_data_username` ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L439)) and `store_data_avatar` ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L495)). Of the two, mousing over an avatar (`store_data_avatar`) appears to work correctly - Profiler returns information about the target blog rather than 'undefined'.

`store_data_username` is called in two ways. It's called directly when mousing over the CSS selectors `.post_info_fence a, .post_info a, a.username` ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L116)). `.post_info_fence a` and `.post_info a` appear in the usernames at the top of posts, and these cases still appears to be working correctly (Profiler returns correct information).

Once a link triggers `store_data_username`, it attempts to get 'popover' information. First it checks for a `data-tumblelog-popover` attribute, and if one exists, [takes the data from it](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L449). This is the case for links at the top of posts, the `.post_info a` ones. If there isn't a `data-tumblelog-popover`, it attempts to get this information [via a HTTP request to the Tumblr API](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L463), at the URL

```javascript
url = "http://www.tumblr.com/svc/tumblelog_popover/" + real_username + "?is_user_mention=false"
```

The problems arise with any links in the body of the post - links in reblog headers, and mentions.

The first problem is that they don't match the three selectors used to call `store_data_username` directly. Instead, there is a catchall function `store_data_username_if_userlink` that attempts to identify any other instances of a link to a blog, checking any other links to see if they end in `.tumblr.com` and calling `store_data_username` if it fits the pattern ([source](https://github.com/new-xkit/XKit/blob/938c0e93b2689b6d6596e4b8c3c9149728186400/Extensions/show_more.js#L421)):

```javascript
store_data_username_if_userlink: function(e) {

	var m_obj = $(e.target);

	if (typeof $(m_obj).attr('href') === "undefined") { return; }

	var m_url = $(m_obj).attr('href').toLowerCase();

	if (m_url.substring(m_url.length - 1) === "/") {m_url = m_url.substring(0, m_url.length - 1); }

	if (m_url.substring(0, 7) === "http://" && m_url.substring(m_url.length - 11) === ".tumblr.com") {
		var m_username = m_url.substring(7, m_url.length - 11);
		XKit.extensions.show_more.store_data_username(e, true, m_username);
	}

},
```

Unfortunately, there aren't very many links that fit this pattern. Links in reblog headers always point to a specific post, for example:

```html
<a class="reblog-tumblelog-name post_info_link" href="http://darmokontheocean.tumblr.com/post/161660731739" data-peepr="{&quot;tumblelog&quot;:&quot;darmokontheocean&quot;, &quot;postId&quot;:&quot;161660731739&quot;}" style="">darmokontheocean</a>
```

And mentions - inline links of the form `@user` - always use a `tumblr.co` link, for example if I type `@canmom` in a post, the resulting URL is https://tmblr.co/mWdhldRm6uUtUj_Ok1pyQ9w rather than https://canmom.tumblr.com:

```html
<a href="https://tmblr.co/mWdhldRm6uUtUj_Ok1pyQ9w" style="">@canmom</a>
```

The first case, links in reblog headers, could be detected by a slightly more complicated function. We want to catch links of the form `http[s]://[username].tumblr.com/something/else`, and `[username]` should not be `www` or it would match all internal Tumblr links in the sidebar etc.. Rather than the relatively simple matching function above, we could use a regular expression to match all suitable URLs. Here's an alternative version of `store_data_username_if_userlink`, using a regex:

```javascript
store_data_username_if_userlink: function(e) {

	var m_obj = $(e.target);

	if (typeof $(m_obj).attr('href') === "undefined") { return; }

	var m_url = $(m_obj).attr('href').toLowerCase();

	var userlink_re = /https?:\/\/(?!www)([a-z]+)\.tumblr\.com[a-z0-9\/-]*/;
	var test = userlink_re.exec(m_url);
	if (test) {
		var m_username = test[1];
		XKit.extensions.show_more.store_data_username(e, true, m_username);
	}

},
```

Making this change in the XKit Editor solved the problem.

The second case, of `tmblr.co` links in mentions, seems like it would be more of a problem. There's no `data-tumblelog-popover`, nor is there any immediate way to deduce the username from the `href`. Somehow Tumblr is able to identify these links. I checked the request data when I moused over this link, and noticed that Tumblr is sending an HTTP request to

```https://www.tumblr.com/svc/tumblelog_popover/m-0lu6vX7CZfGkYoRqi6tzg?is_user_mention=true```

That's something we can easily do by adding an extra clause to `store_data_username` to deal with mentions, and an extra case for `store_data_username_if_userlink` to handle tumblr.co links.

I think I've worked out how to solve the problem while writing this issue, so I'm going to attempt to make the changes and submit a pull request asap.

### PR DESCRIPTION
**Pull request by canmom**
This fixes two cases - mentions, and links to usernames above reblogged text - where User Menus+ was not retrieving the appropriate information about a blog, breaking other extensions such as Profiler.

### CODE REVIEW

**Comment by canmom** [comment]
A friend pointed out that I wasn't capturing URLs with numerals or dashes in my regex, so I made a small commit to address that!

**Comment by nightpool** [comment]
This doesn't seem to work at all for people with custom URLs, is there something more direct we can grab here? How does this work for the native tumblelog_popover call?

I would have to dig more into the exact code flow here to figure out what the bug is, but i'm not sure this solves it in the right way.

**Comment by canmom** [comment]
OK hmm, I didn't think about custom URLs.

In the existing code it tests for:

- links in a `.post_info_fence` block, a `.post_info` block, or with the `username` class, in which case the data is obtained from the `data-tumbelog-popover` attribute
- avatars, which have a `data-tumblelog-popover` attribute
- all links that match the test in the function `store_data_username_if_userlink`, in which case an HTTP request is made to `svc/tumblelog_popover/username?is_mention=false`. Currently this is only testing for links which begin "http://" and end in ".tumblr.com". This fails for reblog headers, which have a popover and no data-tumblelog-popover attribute, but link to something like http(s)://username.tumblr.com/post/2134/title. It also, as you point out, fails for custom URLs. It also fails for URLs with HTTPS enabled, and mentions, which use the tmblr.co format.

The changes in this PR are to the test in the third case. If someone has a custom URL, the first two cases should still be fine. Mentions should also be fine, since they don't use the user's URL (custom or not), but a `tmblr.co` link. That leaves

- inline links
- reblog headers

In the case of reblog headers, I checked a blog with a custom URL, [geographyofrobots.com](http://geographyofrobots.com)/[geographyofrobots.tumblr.com](http://geographyofrobots.tumblr.com). The reblog header goes to the .tumblr.com link, not the custom URL:
picture of reblog header linking to geographyofrobots.tumblr.com
It's only when you click on it that the `.tumblr.com` link gets replaced with the custom URL.

That leaves inline links. To test that, I made a post linking to `http://www.geographyofrobots.com` and moused over it, and it had no popover. So it seems like Tumblr doesn't automatically put a popover on custom URLs.

So I don't think there are any cases where a custom URL would break this actually?

That still leaves the question of if this could be made more efficient by taking advantage of Tumblr's own HTTP requests. Rather than testing on links like this, perhaps we could attach an event to mouseover on the popover element itself? That would require a more significant rewrite, but I would be happy to try.

**Comment by nightpool** [comment]
No, I think that hooking the other info is more of a moonshot. I'm
convinced.

I still have a few style quibbles with the code but I'm on my phone right
now so I can't do a proper review. I'll do it when I get back to my
computer

File: Extensions/show_more.js [hunk 1]
```diff
@@ -424,19 +424,29 @@ XKit.extensions.show_more = new Object({
...
 
-		var m_url = $(m_obj).attr('href').toLowerCase();
+		var m_url = $(m_obj).attr('href');
 
-		if (m_url.substring(m_url.length - 1) === "/") {m_url = m_url.substring(0, m_url.length - 1); }
+		/*Test for a link to a tumblr blog in a form that will have a popover
+		The regex has two capture groups, respectively the username and the path of the linked blog page.
+		https://xyz.tumblr.com links alway have a popover, while https://xyz.tumblr.com/foo/bar links do not have
+		a popover unless they're in a reblog header, in which case they have the post_info_link class.*/
+		var m_test = /https?:\/\/(?!www)([a-z0-9-]+)\.tumblr\.com([a-z0-9/-]*)/.exec(m_url);
```
Comment (hobinjk) [inline Extensions/show_more.js]: I haven't actually seen a blog like this in the wild, but I created a test blog whose URL begins with `www` (`http://wwwtheweb.tumblr.com`) and the regexp doesn't correctly pick up this case.

Comment (canmom) [inline Extensions/show_more.js]: Ah, good point! That can be fixed by changing the regex from `/https?:\/\/(?!www)([a-z0-9-]+)\.tumblr\.com([a-z0-9/-]*)/` to `/https?:\/\/(?!www\.)([a-z0-9-]+)\.tumblr\.com([a-z0-9/-]*)/` which I'll do in just a sec.

**Comment by hobinjk** [comment]
@canmom This looks good to me! The one last thing to do is to increase the value of VERSION in the second line of `show_more.js` from 2.5.5 to 2.5.6. It would also be nice to the next person who reads the code if instead of referring to `m_test[1]` in two places you put `var m_username = m_test[1]` and used `m_username`.

@nightpool How does it look to you?

**Comment by nightpool** [comment]
down for whatever.

**Comment by canmom** [comment]
@hobinjk Done, thank you for checking it over!

**Pull request merged**